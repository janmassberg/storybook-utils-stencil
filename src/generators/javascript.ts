import {
    ArgsType,
    filterEventHandlers,
    filterJsxProperties,
    formatHtml,
    objectToString,
} from "../utils";

export const generateJavascriptCode = (
    component: string,
    props: ArgsType,
    events: ArgsType
): string => {
    if (props.length === 0 && events.length === 0) {
        return "";
    }
    const jsProps =
        props.length > 0
            ? props
                  .map(
                      ([key, value]) =>
                          `element.${key} = ${objectToString(value)};`
                  )
                  .join("\n")
            : "";
    const jsEvents =
        events.length > 0
            ? events
                  .map(
                      ([key, value]) =>
                          `element.addEventListener("${key}", ${objectToString(
                              value
                          )});`
                  )
                  .join("\n")
            : "";
    return `
    const element = document.querySelector("${component}");
    ${jsProps}
    ${jsEvents}
    `;
};

export const generateSourceCodeJavascript = (
    component: string,
    args: any
): string => {
    const eventHandlers = filterEventHandlers(args);
    const jsxProps = filterJsxProperties(args);
    const jsCode = generateJavascriptCode(component, jsxProps, eventHandlers);
    return jsCode === "" ? "" : formatHtml(`<script>${jsCode}</script>`);
};
