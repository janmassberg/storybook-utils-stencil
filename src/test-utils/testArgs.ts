export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: (): void => {
        // Implementation
    },
    onTestEvent: (event: Event): void => {
        // Handle event
        console.log(event.type);
    },
    _slot: "Test children",
};

export const testArgsButton = {
    required: true,
    variant: "primary",
    onClick: (event: Event) => {
        // Handle click event
        console.log(event.target);
    },
    _slot: "Label",
};

export const testArgsComponentGenerator = {
    component: "my-component",
    args: testArgs,
};

export const testCopmonentSlot = `<my-child-component foo-bar="foo">
<my-nested-child-component bar-foo="bar"><span class="test-class">Label</span></my-nested-child-component>
</my-child-component>`;
