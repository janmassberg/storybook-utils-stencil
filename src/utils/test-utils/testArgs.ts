export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: () => {},
    onTestEvent: () => {},
    _children: "Test children",
};
