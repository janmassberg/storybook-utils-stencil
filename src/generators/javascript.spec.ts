import { generateSourceHtmlScript } from "./javascript";
import { testArgsComponentGenerator } from "../test-utils/testArgs";

describe("generateSourceHtmlScript", () => {
    it("should return an empty string if no args or event handlers", () => {
        expect(
            generateSourceHtmlScript({
                ...testArgsComponentGenerator,
                args: {
                    testString: "test string",
                    testNumber: 12345,
                    testTrue: true,
                    testFalse: false,
                    testNull: null,
                    testUndefined: undefined,
                },
            })
        ).toEqual("");
    });

    it("should generate a script tag with prop injection and event handlers", () => {
        expect(
            generateSourceHtmlScript(testArgsComponentGenerator)
        ).toMatchSnapshot();
    });
});
