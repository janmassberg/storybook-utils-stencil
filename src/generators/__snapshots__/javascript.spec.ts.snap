// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateSourceCodeJavascript should generate a script tag with prop injection and event handlers 1`] = `
"<script>
  const element = document.querySelector("my-component");
  element.testArray = [1, 2, 3];
  element.testObject = { foo: "bar" };
  element.testFunction = () => {
    // Implementation
  };
  element.addEventListener("testEvent", (event) => {
    // Handle event
    console.log(event.type);
  });
</script>
"
`;
